File: Submissions/994_Rotting_Oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int, int> > q;
        int m = grid.size(); int n = grid.back().size();
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(grid[i][j] == 2) q.push({i, j});
            }
        }
        
        int time = -1;
        while(!q.empty()){
            int q_size = q.size();
            ++time;
            for(int i=0; i<q_size; ++i){
                auto head = q.front();
                if(head.first > 0 && grid[head.first-1][head.second] == 1) {
                    q.push({head.first-1, head.second});
                    grid[head.first-1][head.second] = 2;
                }
                if(head.first+1 < m && grid[head.first+1][head.second] == 1) {
                    q.push({head.first+1, head.second});
                    grid[head.first+1][head.second] = 2;
                }
                if(head.second > 0 && grid[head.first][head.second-1] == 1) {
                    q.push({head.first, head.second-1});
                    grid[head.first][head.second-1] = 2;
                }
                if(head.second+1 < n && grid[head.first][head.second+1] == 1) {
                    q.push({head.first, head.second+1});
                    grid[head.first][head.second+1] = 2;
                }
                q.pop();
            }
        }
        
        for(int i=0; i<m; ++i){
            for(int j=0; j<n; ++j){
                if(grid[i][j] == 1) return -1;
            }
        }
        
        return max(time, 0);
    }
};

File: Kotlin/994_Rotting_Oranges.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        var num = 0
        var time = -1
        val q: Deque<Pair<Int, Int>> = ArrayDeque()
        
        grid.forEachIndexed { i, row ->
            row.forEachIndexed { j, col ->
                if(col == 1) {
                    ++num
                } else if(col == 2) {
                    q.offer(Pair(i, j))
                }
            }
        }
        
        if(num == 0) return 0

        while (q.isNotEmpty()) {
            ++time
            for (i in 0 until q.size) {
                val head = q.poll()
                val row = head.first
                val col = head.second
                
                if (row >= 1 && grid[row - 1][col] == 1) {
                    grid[row - 1][col] = 2
                    q.offer(Pair(row - 1, col))
                    --num
                }
                if (row < grid.size - 1 && grid[row + 1][col] == 1) {
                    grid[row + 1][col] = 2
                    q.offer(Pair(row + 1, col))
                    --num
                }
                if (col >= 1 && grid[row][col - 1] == 1) {
                    grid[row][col - 1] = 2
                    q.offer(Pair(row, col - 1))
                    --num
                }
                if (col < grid[0].size - 1 && grid[row][col + 1] == 1) {
                    grid[row][col + 1] = 2
                    q.offer(Pair(row, col + 1))
                    --num
                }
            }
        }

        return if (num == 0) time else -1
    }
}